### PR TITLE
Update interop test descriptions.

### PR DESCRIPTION
**Pull request by a11r**
custom_metadata, status_code_and_message and unimplemented_method are now ready for implementation and blocking beta release.

### CODE REVIEW

File: doc/interop-test-descriptions.md [hunk 1]
```diff
@@ -392,22 +392,24 @@ Asserts:
 * clients are free to assert that the response payload body contents are zero
   and comparing the entire response message against a golden response
 
-### Metadata (TODO: fix name)
+### custom_metadata
 
-Status: Not yet implementable
+Status: Ready for implementation, beta requirement.
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: For a completed test, the status can be removed. I had just thrown it there for the tests that had major issues.

File: doc/interop-test-descriptions.md [hunk 2]
```diff
@@ -392,22 +392,24 @@ Asserts:
...
-sent in header and trailer.
+sent as initial-metadata by the client and as both initial- and trailing-metadata
+by the server.
 
 Server features:
 * [UnaryCall][]
+* [FullDuplexCall][]
 * [Compressable Payload][]
 * Ability to receive custom metadata from client in header and send custom data
-  back to client in both header and trailer. (TODO: this is not defined)
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: This still needs to be defined. This should link to a server-side feature in the Server section.

File: doc/interop-test-descriptions.md [hunk 3]
```diff
@@ -392,22 +392,24 @@ Asserts:
...
+  back to client in both header and trailer. 
 
 Procedure:
  1. While sending custom metadata (ascii + binary) in the header, client calls
- UnaryCall with:
+ UnaryCall  with:
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: nit: Remove stray extra space.

File: doc/interop-test-descriptions.md [hunk 4]
```diff
@@ -418,21 +420,31 @@ Procedure:
       }
     }
     ```
+The client attaches custom metadat with the following keys:
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: s/metadat/metadata/

File: doc/interop-test-descriptions.md [hunk 5]
```diff
@@ -418,21 +420,31 @@ Procedure:
...
-* custom metadata is echoed back in the response trailer.
+* metadata with key `"x-grpc-test-echo-initial"` is received in the initial metadata.
+* metadata with key `"x-grpc-test-echo-trailing"` is received in the trailing metadata.
+
+
 
 ### status_code_and_message
 
-Status: Not yet implementable
+Status: Ready for implementation, beta requirement.
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: This is not implementable, because server-side needs to be defined.

File: doc/interop-test-descriptions.md [hunk 6]
```diff
@@ -452,10 +466,9 @@ Asserts:
 
 ### unimplemented_method
 
-Status: Not yet implementable
+Status: Ready for implementation. Blocking beta.
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: This is not implementable because UnimplementedCall is not defined. It seems we should have a second service in test.proto that has the unimplemented call.

File: doc/interop-test-descriptions.md (hunk 1 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: Done.

File: doc/interop-test-descriptions.md (hunk 3 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: done

File: doc/interop-test-descriptions.md (hunk 4 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: done

File: doc/interop-test-descriptions.md (hunk 5 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: Added a spec for the required server-side behavior as "Echo Status".

File: doc/interop-test-descriptions.md (hunk 6 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: Agreed. Will do that as a separate PR since it will need to trigger all testing.

**Comment by a11r** [comment]
Thanks for the review. I have addressed all comments. PTAL.

File: doc/interop-test-descriptions.md [hunk 7]
```diff
@@ -681,6 +695,14 @@ When the client requests COMPRESSABLE payload, the response includes a payload
 of the size requested containing all zeros and the payload type is
 COMPRESSABLE.
 
+### Echo Status
+[Echo Status]: #echo-status
+When the client sends a response_status in the request payload, the server returns
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: `response_status` does not exist in messages.proto. This should specify which messages to look for response_status. What should the behavior be for a StreamingInputCallRequest? Should the server use the last received status?

File: doc/interop-test-descriptions.md (hunk 6 above)
Comment (ejona86) [inline doc/interop-test-descriptions.md]: Then let's leave the status as "Not yet implementable" here.

File: doc/interop-test-descriptions.md (hunk 1 above)
Comment (ejona86) [inline doc/interop-test-descriptions.md]: That is different than we've done in the past. We've tracked implementation status separately. Is there a reason to doing that differently now? If so, then multiple pre-existing tests need to a status added because they aren't implemented in all languages.

File: doc/interop-test-descriptions.md (hunk 2 above)
Comment (ejona86) [inline doc/interop-test-descriptions.md]: This still needs to be defined.

File: doc/interop-test-descriptions.md (hunk 7 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: response_status is now in messages.proto. Clients setting response_status in StreamingInputCallRequest should set the same value in all messages in the stream. And the server should be free to use this value from any message in the client->server stream. Using this value from the last received message seems a reasonable server impl.

**Comment by a11r** [comment]
Foirgot to update this thread. PTAL.

File: doc/interop-test-descriptions.md (hunk 7 above)
Comment (ejona86) [inline doc/interop-test-descriptions.md]: Based on the text written, I would never expect the server to experience more than one message with response_status. Instead, I would expect the server to immediately close the stream when it gets the first response_status and so would ignore any further messages. Is that not the intention?

File: doc/interop-test-descriptions.md (hunk 2 above)
Comment (ejona86) [inline doc/interop-test-descriptions.md]: This still needs to be defined.

File: doc/interop-test-descriptions.md (hunk 7 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: Your interpretation is correct. I am happy to add any further clarifications needed.

File: doc/interop-test-descriptions.md [hunk 8]
```diff
@@ -418,21 +420,32 @@ Procedure:
       }
     }
     ```
+The client attaches custom metadata with the following keys:
+    ```
+    "x-grpc-test-echo-initial"
+    or
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: I don't understand this "or". What is the decision based on? If the client only attaches one, then how are both asserted?

File: doc/interop-test-descriptions.md (hunk 6 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: This is implementable now.

File: doc/interop-test-descriptions.md (hunk 8 above)
Comment (a11r) [inline doc/interop-test-descriptions.md]: The two keys refer to two different cases. In the first case the client is asking for initial metadata in the response stream , in the second case the client is asking for trailing metadata.

File: doc/interop-test-descriptions.md (hunk 6 above)
Comment (ejona86) [inline doc/interop-test-descriptions.md]: It is not implementable. The client is expected to do things it can't do (UnimplementedCall accepts an Empty).

We should probably also specify the service is UnimplementedService.

File: doc/interop-test-descriptions.md (hunk 8 above)
Comment (ejona86) [inline doc/interop-test-descriptions.md]: So the client is supposed to issue two calls? Or should the client issue one call with both?

I guess also, what should the values of the keys be?

**Comment by a11r** [comment]
Confirming that I am the author.

File: doc/interop-test-descriptions.md [hunk 9]
```diff
@@ -509,21 +508,29 @@ Procedure:
       }
     }
     ```
+The client attaches custom metadata with the following keys and values:
+    ```
+    key: "x-grpc-test-echo-initial", value: "test_initial_metadata_value"
+    key: "x-grpc-test-echo-trailing", value: 0xababab
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: `0xababab` isn't ASCII and so isn't valid for that key. Were you wanting it to be a binary key?

File: doc/interop-test-descriptions.md [hunk 10]
```diff
@@ -543,21 +552,15 @@ Asserts:
...
-Status: Not yet implementable
+Status: Ready for implementation. Blocking beta.
 
-This test verifies calling unimplemented RPC method returns unimplemented
-status.
+This test verifies calling unimplemented RPC method returns the UNIMPLEMENTED status code.
 
 Procedure:
-* Client calls UnimplementedCall with:
+* Client calls UnimplementedCall with an empty request (defined as `grpc.testing.Empty`):
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: Since this isn't the normal test service, it does seem good to note the service name.

**Comment by ctiller** [comment]
Ping.

**Comment by a11r** [comment]
Addressed all comments. PTAL.

File: doc/interop-test-descriptions.md [hunk 11]
```diff
@@ -509,21 +508,29 @@ Procedure:
...
+    key: "x-grpc-test-echo-trailing-bin", value: 0xababab
+    ```
+ 2. Client repeats step 1. with FullDuplexCall instead of UnaryCall.
 
 Asserts:
 * call was successful
-* custom metadata is echoed back in the response header.
-* custom metadata is echoed back in the response trailer.
+* metadata with key `"x-grpc-test-echo-initial"` and value `"test_initial_metadata_value"`is received in the initial metadata.
+* metadata with key `"x-grpc-test-echo-trailing-bin"` and value `0xababab` is received in the trailing metadata.
```
Comment (ejona86) [inline doc/interop-test-descriptions.md]: This won't work, because the server-side Echo Metadata feature echos x-grpc-test-echo-trailing, not x-grpc-test-echo-trailing-bin. The server feature needs to be updated in some way. For example, if it also echoed x-grpc-test-echo-trailing-bin and x-grpc-test-echo-trailing-bin then it would work.

Comment (a11r) [inline doc/interop-test-descriptions.md]: :) Updated the server feature as well. PTAL.
